<a href="https://colab.research.google.com/github/dgullate/Curso-IE/blob/master/Week_1/pandas_completo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas
![alt text](https://pandas.pydata.org/_static/pandas_logo.png)

Pandas is a python library with high level data structures that have many functionalities and are easy to use for the purpose of data analytics.
The main objects are called `dataframes`, which are 2-dim data structures similar to a table (and to dataframes in R).

In [ ]:
!git clone https://github.com/dgullate/Curso-IE.git

In [1]:
import pandas as pd #Así importamos la librería pandas
import numpy as np

## Creating dataframes

We can create a dataframe from various Series, each of them will be a register in our table. For instance, let's create a table with 3 registries whose columns are `Cost`, `Item Purchased` y `Name`:

In [20]:

purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])

As we see, it is enough to define a Series object for each entry in our table. Moreover, the data types must be consistent on each Series object, so that the table get the right formatting across columns. The `index` argument specifies identifiers for each item. By defect, an index is an integer number, but in this case we used strings.

In [21]:
df.head()

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

We can select a subset of elements knowing its index, using `loc`.

In [22]:
df.loc['Store 1']

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5

If moreover we want to select a specific columns, we can do so separating with a comma.

In [23]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

Even more ! these arguments can be both lists.

In [24]:
df.loc[['Store 1', 'Store 2'], ['Cost','Name']]

Cost   Name
Store 1  22.5  Chris
Store 1   2.5  Kevyn
Store 2   5.0  Vinod

If instead of accesing via the index, we would like to access by the row or column positions, we can do so using `iloc`. As an example, the following command selects the first 2 rows and all columns.


In [25]:
df.iloc[:3, :]

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

<br>
If we want to erase some item knowing its index, we can do so using `drop`

In [26]:
df.drop('Store 1')

Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0

In [27]:
df

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

What happened ? Didn't we erase it ? The thing is `drop` by default creates a new copy of the dataframe, instead of assigning a new variable `df2 = df.drop(...)` we can specify that the updated dataframe overwrites the same object in memory by using the flag `inplace=True` .

In [28]:
df.drop('Store 1', inplace=True)
df

Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0

## Loading dataframes

Most of the times we will not work with df in this manner, as they will be lying on some file in a hard disk or web server. Pandas has functions that are able to read and load the contents of these files, and create a dataframe object from them.
All of these functions start like `pd.read_...` and there are versions ofr file types in excel, html, json, etc. We will use the most common function  `pd.read_csv`, that reads .csv files(every item int he dataframe is a line in the csv files, and vaues are usually comma separated).

Next, we load the data contained in the file `tips.csv` in a  `pandas`  `dataframe` . This database contains information on the cost of meals and tips in restaurants. It has the following variables:

* `total_bill`: total proce of the meal.
* `tip`: amount given as tip.
* `sex`: Gender of the paying person (Female/Male).
* `smoker`: Categorical variable indicating whether the payer is a smoker (yes/no).
* `day`: Day of thre week.
* `time`: Categorical variable indicating if it was lunch or dinner.
* `size`: Number of people at the table

In [29]:
#dir='/content/Curso-IE/Week_1/data/tips.csv'
dir='data/tips.csv'
df = pd.read_csv(dir)

Besides the name of the file, a necessary argument, `read_csv` has other optional args. The most common are:
    
* nrows: only reads a certain number of lines (useful to test before we load a very large file).
* usecols: the dataframe will only red these columns.
* dtype: specifies the data type for each column (by defect, pandas tries to infer the type automatically by looking at the data).

## Exploratory analysis

Once we have loaded the data, we can look at them with the following four functions: `head` and `tail` to observe the first and last lines; para  `describe` to obtain basic statistic summary and  `info` to obtain the data types of each columns.

In [30]:
df.head(10) #in paranthesis, th enumber of files to be displayed
            # by default, that's 5.

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
6        8.77  2.00    Male     No  Sun  Dinner     2
7       26.88  3.12    Male     No  Sun  Dinner     4
8       15.04  1.96    Male     No  Sun  Dinner     2
9       14.78  3.23    Male     No  Sun  Dinner     2

In [31]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [32]:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

For columns with numerica data, `describe` gives information on the number of missing values, the mean value, standard deviation and some quantiles.
If the table also contains columns with categorical data,  `describe` by default only gives info in the numerical ones. If want info on the other columns, we must write:

In [33]:
df.describe(include='all')

total_bill         tip   sex smoker  day    time        size
count   244.000000  244.000000   244    244  244     244  244.000000
unique         NaN         NaN     2      2    4       2         NaN
top            NaN         NaN  Male     No  Sat  Dinner         NaN
freq           NaN         NaN   157    151   87     176         NaN
mean     19.785943    2.998279   NaN    NaN  NaN     NaN    2.569672
std       8.902412    1.383638   NaN    NaN  NaN     NaN    0.951100
min       3.070000    1.000000   NaN    NaN  NaN     NaN    1.000000
25%      13.347500    2.000000   NaN    NaN  NaN     NaN    2.000000
50%      17.795000    2.900000   NaN    NaN  NaN     NaN    2.000000
75%      24.127500    3.562500   NaN    NaN  NaN     NaN    3.000000
max      50.810000   10.000000   NaN    NaN  NaN     NaN    6.000000

For categorical variables, we also see information on the number of unique values, the most frequent value and its frequency. For instance, in the gender case, there are 244 entries, the most frquent value is Male (157 out of 244), there's only 2 unique values (so we assume the other value is Female)

In [34]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
total_bill    244 non-null float64
tip           244 non-null float64
sex           244 non-null object
smoker        244 non-null object
day           244 non-null object
time          244 non-null object
size          244 non-null int64
dtypes: float64(2), int64(1), object(4)
memory usage: 13.4+ KB


## Querying dataframes

So far we have seen how to access certain items and/or certain columns in our dataframe. Let's see now other ways to make more dynamic queries.

A quick way to access a whole column is:

In [35]:
df['tip'].head()

0    1.01
1    1.66
2    3.50
3    3.31
4    3.61
Name: tip, dtype: float64

We could also do

In [36]:
df.tip.head()

0    1.01
1    1.66
2    3.50
3    3.31
4    3.61
Name: tip, dtype: float64

As pandas is built on numpy, we can use a very similar syntax.

In [37]:
df['size'].median(), df['size'].mean()

(2.0, 2.569672131147541)

In [38]:
df['size'] > df['size'].median()

0      False
1       True
2       True
3      False
4       True
5       True
6      False
7       True
8      False
9      False
10     False
11      True
12     False
13      True
14     False
15     False
16      True
17      True
18      True
19      True
20     False
21     False
22     False
23      True
24     False
25      True
26     False
27     False
28     False
29     False
       ...  
214     True
215    False
216     True
217    False
218    False
219     True
220    False
221    False
222    False
223     True
224    False
225    False
226    False
227     True
228    False
229    False
230     True
231     True
232    False
233    False
234    False
235    False
236    False
237    False
238     True
239     True
240    False
241    False
242    False
243    False
Name: size, Length: 244, dtype: bool

The previous command has generated a Boolean array, so we can use it as a mask to select rows in our dataframe !

In [39]:
large_tables = df[ df['size'] > df['size'].median() ]
large_tables.head()

total_bill   tip     sex smoker  day    time  size
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
7       26.88  3.12    Male     No  Sun  Dinner     4

We selected those tables that are larger than the median value, but we see they keep the original index. If we want to reset the index we must do

In [40]:
large_tables.reset_index(inplace=True, drop=True)
large_tables.head()

total_bill   tip     sex smoker  day    time  size
0       10.34  1.66    Male     No  Sun  Dinner     3
1       21.01  3.50    Male     No  Sun  Dinner     3
2       24.59  3.61  Female     No  Sun  Dinner     4
3       25.29  4.71    Male     No  Sun  Dinner     4
4       26.88  3.12    Male     No  Sun  Dinner     4

The criteria to filter and select rows can be as complex as we wish, using logic operators.

In [41]:
df[ (df['size'] > df['size'].median()) & (df['sex'] == 'Female') ].head()

total_bill   tip     sex smoker  day    time  size
4        24.59  3.61  Female     No  Sun  Dinner     4
11       35.26  5.00  Female     No  Sun  Dinner     4
16       10.33  1.67  Female     No  Sun  Dinner     3
18       16.97  3.50  Female     No  Sun  Dinner     3
33       20.69  2.45  Female     No  Sat  Dinner     4


##    Exercise 1
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 


Load the table `olympics.csv`. If you look at the file, you must skip the first row (use `skiprows`). Moreover, we can tell pandas to use the first column as index (index_col=0). Let's to some basic exploratory analysis of the data.

In [92]:
#dir='/content/Curso-IE/Week_1/data/olympics.csv'
dir='data/olympics.csv'
olympics = pd.read_csv(dir, skiprows=1, index_col=0)

To help us, run the following code to rename the columns.

In [93]:
for col in olympics.columns:
    if col[:2]=='01':
        olympics.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        olympics.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        olympics.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        olympics.rename(columns={col:'#' + col[1:]}, inplace=True) 
        
names_ids = olympics.index.str.split('\s\(') # split the index by '('

olympics.index = names_ids.str[0] # the [0] element is the country name (new index) 
olympics['ID'] = names_ids.str[1].str[:3] # the [1] element is the abbreviation or ID (take first 3 characters from that)
        
olympics =  olympics.drop('Totals')
olympics.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

In [44]:
olympics.describe()

# Summer        Gold      Silver      Bronze        Total  \
count  146.000000  146.000000  146.000000  146.000000   146.000000   
mean    13.383562   32.938356   32.705479   35.136986   100.780822   
std      7.006657   96.745184   80.648421   77.118416   252.700436   
min      1.000000    0.000000    0.000000    0.000000     0.000000   
25%      8.000000    0.000000    1.000000    1.000000     2.000000   
50%     13.000000    3.000000    4.000000    6.000000    12.000000   
75%     18.000000   23.000000   26.750000   28.750000    85.250000   
max     27.000000  976.000000  757.000000  666.000000  2399.000000   

         # Winter      Gold.1    Silver.1    Bronze.1     Total.1     # Games  \
count  146.000000  146.000000  146.000000  146.000000  146.000000  146.000000   
mean     6.595890    6.568493    6.561644    6.493151   19.623288   19.979452   
std      7.349012   18.975086   18.606996   17.499654   54.640914   13.084062   
min      0.000000    0.000000    0.000000    0.000000    0.000000    1.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000   11.000000   
50%      5.000000    0.000000    0.000000    0.000000    0.000000   15.000000   
75%     10.000000    0.750000    1.750000    1.000000    4.750000   26.000000   
max     22.000000  118.000000  111.000000  100.000000  329.000000   49.000000   

            Gold.2    Silver.2    Bronze.2  Combined total  
count   146.000000  146.000000  146.000000      146.000000  
mean     39.506849   39.267123   41.630137      120.404110  
std     109.932471   93.984393   89.826117      291.988913  
min       0.000000    0.000000    0.000000        1.000000  
25%       0.000000    1.000000    1.000000        2.250000  
50%       3.000000    4.000000    6.500000       12.000000  
75%      24.500000   28.000000   29.000000       87.750000  
max    1072.000000  859.000000  750.000000     2681.000000

¿Which country obtained the highest number of gold medals in the Summer Olympic Games ?

In [45]:
olympics[olympics['Gold']==max(olympics['Gold'])]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
United States        26   976     757     666   2399        22      96   

               Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
United States       102        84      282       48    1072       859   

               Bronze.2  Combined total   ID  
United States       750            2681  USA

This is actually a shortcut:

In [46]:
olympics.Gold.idxmax()

'United States'

List the countries that have obtained more medals (each medal counts the same) in the winter that in the summer games.

In [47]:
## your answer here
list(olympics[olympics['Total'] < olympics['Total.1']].index)

['Austria', 'Liechtenstein', 'Norway']

## Creating new columns and applying functions on columns

To add a new column to our table, it is as simple as assigning a new column value.

In [48]:
df['new_col'] = None
df.head()

total_bill   tip     sex smoker  day    time  size new_col
0       16.99  1.01  Female     No  Sun  Dinner     2    None
1       10.34  1.66    Male     No  Sun  Dinner     3    None
2       21.01  3.50    Male     No  Sun  Dinner     3    None
3       23.68  3.31    Male     No  Sun  Dinner     2    None
4       24.59  3.61  Female     No  Sun  Dinner     4    None

We can also create columns using numpy functions. For instance, we will fill the new column with random numbers sampled from a normal distribution.

In [49]:
df['new_col'] = np.random.randn(len(df))
df.head()

total_bill   tip     sex smoker  day    time  size   new_col
0       16.99  1.01  Female     No  Sun  Dinner     2  0.464786
1       10.34  1.66    Male     No  Sun  Dinner     3 -0.934138
2       21.01  3.50    Male     No  Sun  Dinner     3  2.130752
3       23.68  3.31    Male     No  Sun  Dinner     2 -0.866834
4       24.59  3.61  Female     No  Sun  Dinner     4  0.160842

We can also combine other columns of the dataframe.

In [50]:
del df['new_col']
df['total_bill_rand'] = np.random.randn(len(df)) + df['total_bill']
df.head()

total_bill   tip     sex smoker  day    time  size  total_bill_rand
0       16.99  1.01  Female     No  Sun  Dinner     2        19.229865
1       10.34  1.66    Male     No  Sun  Dinner     3        11.235986
2       21.01  3.50    Male     No  Sun  Dinner     3        21.413654
3       23.68  3.31    Male     No  Sun  Dinner     2        24.105713
4       24.59  3.61  Female     No  Sun  Dinner     4        23.437885

A very important method in pandas is `apply`, which allows to apply any function( from numpy, pandas or defined by us) on a given column or list of columns. 

For instance, we can calculate the square root of the columns total_bill, tip y total_bill_rand:

In [51]:
df[['total_bill', 'tip', 'total_bill_rand']].apply(np.sqrt).head()

total_bill       tip  total_bill_rand
0    4.121893  1.004988         4.385187
1    3.215587  1.288410         3.352012
2    4.583667  1.870829         4.627489
3    4.866210  1.819341         4.909757
4    4.958831  1.900000         4.841269

In [52]:
df.apply(np.max)

total_bill           50.81
tip                     10
sex                   Male
smoker                 Yes
day                   Thur
time                 Lunch
size                     6
total_bill_rand    50.1505
dtype: object

LEt's see another example with a user defined function_

In [53]:
def small_filter(x, y):   
    x[x<y] = 0
    return x

In [54]:
df[['total_bill']].apply(small_filter, args=[df['total_bill'].mean()]).head()

total_bill
0        0.00
1        0.00
2       21.01
3       23.68
4       24.59

Another way to do it is to use `where`:  if the condition is True, the value is unchanged, but if it is False, the argument is replaced with the alternative argument.

In [55]:
mask=df['total_bill']>df['total_bill'].mean()
df['total_bill'].where(mask,0).head()

0     0.00
1     0.00
2    21.01
3    23.68
4    24.59
Name: total_bill, dtype: float64

### Exercise

We work on the olympics dataset. Get the countries that have obtained more points in winter than summer games. Points are calculated in the following way: bronze 1, silver 2, gold 3.

In [56]:
olympics['points_summer'] = 3*olympics['Gold'] + 2*olympics['Silver'] + olympics['Bronze']
olympics['points_winter'] = 3*olympics['Gold.1'] + 2*olympics['Silver.1'] + olympics['Bronze.1']

In [57]:
olympics[olympics['points_summer'] < olympics['points_winter']]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Austria              26    18      33      35     86        22      59   
Liechtenstein        16     0       0       0      0        18       2   
Norway               24    56      49      43    148        22     118   

               Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Austria              78        81      218       48      77       111   
Liechtenstein         2         5        9       34       2         2   
Norway              111       100      329       46     174       160   

               Bronze.2  Combined total   ID  points_summer  points_winter  
Austria             116             304  AUT            155            414  
Liechtenstein         5               9  LIE              0             15  
Norway              143             477  NOR            309            676

¿Which country has the smallest relative diference between summer and winter points, i.e. compute for every country
$$
dif\_rel = \frac{|points\_ summer - points\_winter|}{points\_summer + points\_winter} 
$$

In [58]:
import numpy as np
olympics['dif_rel'] = np.abs(olympics['points_summer'] - olympics['points_winter'])/(olympics['points_summer'] + olympics['points_winter'])

In [59]:
olympics[olympics['dif_rel'] == min(olympics['dif_rel'])]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
Luxembourg        22     1       1       0      2         8       0         2   

            Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Luxembourg         0        2       30       1         3         0   

            Combined total   ID  points_summer  points_winter   dif_rel  
Luxembourg               4  LUX              5              4  0.111111

## Merging dataframes

It often happens that relevant information is spread across different tables. For this purpose, pandas has the `merge` operation, that allows to bring together items coming from different tables (a bit like a join in SQL). Let us start with some example data.

In [60]:
staff_df = pd.DataFrame([{'Name': 'Kelly', 'Role': 'Director of HR'},
                         {'Name': 'Sally', 'Role': 'Course liasion'},
                         {'Name': 'James', 'Role': 'Grader'}])
staff_df = staff_df.set_index('Name')
student_df = pd.DataFrame([{'Name': 'James', 'School': 'Business'},
                           {'Name': 'Mike', 'School': 'Law'},
                           {'Name': 'Sally', 'School': 'Engineering'}])
student_df = student_df.set_index('Name')
print(staff_df.head())
print()
print(student_df.head())

                 Role
Name                 
Kelly  Director of HR
Sally  Course liasion
James          Grader

            School
Name              
James     Business
Mike           Law
Sally  Engineering


`merge` needs to input 2 tables, the necessary argument `how`, and one column on each table that will be used to establish the correspondence (by default, the index will be taken unless otherwise specified.

Let us see the different options that the `how` argument can take.

<img src="https://media.geeksforgeeks.org/wp-content/uploads/joinimages.png"  width="400"/> 

With `outer` the combined table has all the items from the left table (`staff_df`) and the right table (`school_df`) even if there is no matching. For instance, we see that James is Grader on the left table, and he belongs to the Business Faculty, so both values appear in the combined table. But Kelly only appears in the Role table, so in the combined table information on her Faculty is missing (NaN):

In [61]:
pd.merge(staff_df, student_df, how='outer', left_index=True, right_index=True)

Role       School
Name                              
James          Grader     Business
Kelly  Director of HR          NaN
Mike              NaN          Law
Sally  Course liasion  Engineering

In [62]:
pd.merge(staff_df, student_df, how='inner', left_index=True, right_index=True)

Role       School
Name                              
Sally  Course liasion  Engineering
James          Grader     Business

`left` only takes the rows that appear in the left table, and tries to bring in information on the other columns from the right table (whenever possible):

In [63]:
pd.merge(staff_df, student_df, how='left', left_index=True, right_index=True)

Role       School
Name                              
Kelly  Director of HR          NaN
Sally  Course liasion  Engineering
James          Grader     Business

`right` works the same as `left`:

In [64]:
pd.merge(staff_df, student_df, how='right', left_index=True, right_index=True)

Role       School
Name                              
James          Grader     Business
Mike              NaN          Law
Sally  Course liasion  Engineering

Now we will reset the indices and perform a `merge` specifying whicvh columns we want to use as key in each table:

In [65]:
staff_df = staff_df.reset_index()
student_df = student_df.reset_index()
pd.merge(staff_df, student_df, how='left', left_on='Name', right_on='Name')

Name            Role       School
0  Kelly  Director of HR          NaN
1  Sally  Course liasion  Engineering
2  James          Grader     Business

One small thing: note that if other than the key column, there are other columns on the two tables that have the same name, when merging they will be given different names in order to avoid collisions.

In [68]:
staff_df = pd.DataFrame([{'Name': 'Kelly', 'Role': 'Director of HR', 'Location': 'State Street'},
                         {'Name': 'Sally', 'Role': 'Course liasion', 'Location': 'Washington Avenue'},
                         {'Name': 'James', 'Role': 'Grader', 'Location': 'Washington Avenue'}])
student_df = pd.DataFrame([{'Name': 'James', 'School': 'Business', 'Location': '1024 Billiard Avenue'},
                           {'Name': 'Mike', 'School': 'Law', 'Location': 'Fraternity House #22'},
                           {'Name': 'Sally', 'School': 'Engineering', 'Location': '512 Wilson Crescent'}])

print(staff_df.head())
print()
print(student_df.head())

            Location   Name            Role
0       State Street  Kelly  Director of HR
1  Washington Avenue  Sally  Course liasion
2  Washington Avenue  James          Grader

               Location   Name       School
0  1024 Billiard Avenue  James     Business
1  Fraternity House #22   Mike          Law
2   512 Wilson Crescent  Sally  Engineering


In [69]:
pd.merge(staff_df, student_df, how='left', left_on='Name', right_on='Name')

Location_x   Name            Role            Location_y       School
0       State Street  Kelly  Director of HR                   NaN          NaN
1  Washington Avenue  Sally  Course liasion   512 Wilson Crescent  Engineering
2  Washington Avenue  James          Grader  1024 Billiard Avenue     Business

We can also merge using more than one column as keys for matching:

In [72]:
staff_df = pd.DataFrame([{'First Name': 'Kelly', 'Last Name': 'Desjardins', 'Role': 'Director of HR'},
                         {'First Name': 'Sally', 'Last Name': 'Brooks', 'Role': 'Course liasion'},
                         {'First Name': 'James', 'Last Name': 'Wilde', 'Role': 'Grader'}])
student_df = pd.DataFrame([{'First Name': 'James', 'Last Name': 'Hammond', 'School': 'Business'},
                           {'First Name': 'Mike', 'Last Name': 'Smith', 'School': 'Law'},
                           {'First Name': 'Sally', 'Last Name': 'Brooks', 'School': 'Engineering'}])
print(staff_df)
print()
print(student_df)


  First Name   Last Name            Role
0      Kelly  Desjardins  Director of HR
1      Sally      Brooks  Course liasion
2      James       Wilde          Grader

  First Name Last Name       School
0      James   Hammond     Business
1       Mike     Smith          Law
2      Sally    Brooks  Engineering


In [73]:
pd.merge(staff_df, student_df, how='inner', left_on=['First Name','Last Name'], right_on=['First Name','Last Name'])


First Name Last Name            Role       School
0      Sally    Brooks  Course liasion  Engineering

## Exercise 2 
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 

Let's go back to the Olympics. Load the table  `population.csv` that provides the population of each country across a number of years. To make things simple, we will only keep the populations in year 2016 (you need to select them).
Once you have done that, combine the tables olympics and population (such that there cannot be any missing values). Make sure in the combined table we have at least the following 3 columns: name of the country, number of gold medals and population.

Which country has the higher ratio of gold medals per capita ?


In [88]:
#load the file
dir_pop ='data/population.csv'
dir_oly ='data/olympics.csv'
# dir='/content/Curso-IE/Week_1/data/population.csv'
pop = pd.read_csv(dir_pop)
olympics=pd.read_csv(dir_oly)

In [89]:
pop.tail()

Country Name Country Code  Year       Value
14880     Zimbabwe          ZWE  2012  14710826.0
14881     Zimbabwe          ZWE  2013  15054506.0
14882     Zimbabwe          ZWE  2014  15411675.0
14883     Zimbabwe          ZWE  2015  15777451.0
14884     Zimbabwe          ZWE  2016  16150362.0

In [90]:
#select only the population of year 2016
pop = pop[pop['Year'] == 2016]
pop.Value=pop.Value.astype('int')
del pop['Country Code']
pop.head()

Country Name  Year       Value
56                       Arab World  2016   406452690
113          Caribbean small states  2016     7245472
170  Central Europe and the Baltics  2016   102974082
227      Early-demographic dividend  2016  3170542188
284             East Asia & Pacific  2016  2296786207

In [94]:
olympics.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

In [95]:
#reset the index in the olympics_df
olympics = olympics.reset_index()
olympics.head()

index  # Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
0  Afghanistan        13     0       0       2      2         0       0   
1      Algeria        12     5       2       8     15         3       0   
2    Argentina        23    18      24      28     70        18       0   
3      Armenia         5     1       2       9     12         6       0   
4  Australasia         2     3       4       5     12         0       0   

   Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
0         0         0        0       13       0         0         2   
1         0         0        0       15       5         2         8   
2         0         0        0       41      18        24        28   
3         0         0        0       11       1         2         9   
4         0         0        0        2       3         4         5   

   Combined total   ID  
0               2  AFG  
1              15  ALG  
2              70  ARG  
3              12  ARM  
4              12  ANZ

In [96]:
oly_pop = pd.merge(olympics, pop, how='inner', left_on='index', right_on='Country Name')
oly_pop.head()

index  # Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
0  Afghanistan        13     0       0       2      2         0       0   
1      Algeria        12     5       2       8     15         3       0   
2    Argentina        23    18      24      28     70        18       0   
3      Armenia         5     1       2       9     12         6       0   
4    Australia        25   139     152     177    468        18       5   

   Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
0         0         0        0       13       0         0         2   
1         0         0        0       15       5         2         8   
2         0         0        0       41      18        24        28   
3         0         0        0       11       1         2         9   
4         3         4       12       43     144       155       181   

   Combined total   ID Country Name  Year     Value  
0               2  AFG  Afghanistan  2016  34656032  
1              15  ALG      Algeria  2016  40606052  
2              70  ARG    Argentina  2016  43847430  
3              12  ARM      Armenia  2016   2924816  
4             480  AUS    Australia  2016  24127159

In [97]:
# Pick only the necessary columns
oly_pop=oly_pop[['Country Name','Gold','Value']]
oly_pop.head()

Country Name  Gold     Value
0  Afghanistan     0  34656032
1      Algeria     5  40606052
2    Argentina    18  43847430
3      Armenia     1   2924816
4    Australia   139  24127159

In [99]:
oly_pop['ratio'] = oly_pop['Gold'] / oly_pop['Value']
oly_pop.head()

Country Name  Gold     Value         ratio
0  Afghanistan     0  34656032  0.000000e+00
1      Algeria     5  40606052  1.231344e-07
2    Argentina    18  43847430  4.105144e-07
3      Armenia     1   2924816  3.419018e-07
4    Australia   139  24127159  5.761142e-06

In [100]:
oly_pop[['Country Name', 'ratio']].sort_values(by='ratio', ascending=False).reset_index(drop=True).head(10)

Country Name     ratio
0      Finland  0.000018
1      Hungary  0.000017
2       Sweden  0.000014
3       Norway  0.000011
4      Grenada  0.000009
5  New Zealand  0.000009
6      Denmark  0.000008
7     Bulgaria  0.000007
8      Estonia  0.000007
9         Cuba  0.000006

Now do the same thing as before, but rank the 10 countries that have the largest ratio of gold medals to GDP. Can you guess before which will be the first ?

In [ ]:
gdp = pd.read_csv('/content/Curso-IE/Week_1/data/gdp.csv')

In [ ]:
gdp = gdp[gdp['Year'] == 2015]
gdp_oly_pop= pd.merge(oly_pop, gdp, how='inner', left_on='index', right_on='Country Name')
gdp_oly_pop['gold2gdp_ratio']=gdp_oly_pop['Gold']/gdp_oly_pop['Value_y']
gdp_oly_pop[['Country Name_x', 'gold2gdp_ratio']].sort_values(by='gold2gdp_ratio', ascending=False).reset_index(drop=True).head(10)

## Grouping values in dataframes
A very important function in pandas is `groupby`, that we can use for the following purposes

* split the table in groups accoridng to some criteria
* apply some operation to some group independently
* combine the results into a new table with aggregated data

Let us get started with creating groups. The typical procedure is to split according to the values of some categorical variables (gender, day of the week, etc.)

Let us do some examples on the table `tips.csv`

In [101]:
for name, group in df.groupby(by=['sex']):
    print(name)
    print(group)

Female
     total_bill   tip     sex smoker   day    time  size  total_bill_rand
0         16.99  1.01  Female     No   Sun  Dinner     2        19.229865
4         24.59  3.61  Female     No   Sun  Dinner     4        23.437885
11        35.26  5.00  Female     No   Sun  Dinner     4        36.013951
14        14.83  3.02  Female     No   Sun  Dinner     2        14.169146
16        10.33  1.67  Female     No   Sun  Dinner     3         9.978721
18        16.97  3.50  Female     No   Sun  Dinner     3        19.116772
21        20.29  2.75  Female     No   Sat  Dinner     2        18.853396
22        15.77  2.23  Female     No   Sat  Dinner     2        16.937855
29        19.65  3.00  Female     No   Sat  Dinner     2        17.987859
32        15.06  3.00  Female     No   Sat  Dinner     2        15.914629
33        20.69  2.45  Female     No   Sat  Dinner     4        21.351903
37        16.93  3.07  Female     No   Sat  Dinner     3        17.675599
51        10.29  2.60  Female  

We can select more than one variable to make groups. For instance, if we group by sex and size, since there are 2 values for sex and 6 values for size, there will be 12 groups created.

In [104]:
for name, group in df.groupby(by=['sex', 'size']):
    print(name)
    print(group)


     total_bill   tip     sex smoker   day    time  size  total_bill_rand
67         3.07  1.00  Female    Yes   Sat  Dinner     1         3.364872
82        10.07  1.83  Female     No  Thur   Lunch     1         9.077067
111        7.25  1.00  Female     No   Sat  Dinner     1         5.507585
     total_bill   tip     sex smoker   day    time  size  total_bill_rand
0         16.99  1.01  Female     No   Sun  Dinner     2        19.229865
14        14.83  3.02  Female     No   Sun  Dinner     2        14.169146
21        20.29  2.75  Female     No   Sat  Dinner     2        18.853396
22        15.77  2.23  Female     No   Sat  Dinner     2        16.937855
29        19.65  3.00  Female     No   Sat  Dinner     2        17.987859
32        15.06  3.00  Female     No   Sat  Dinner     2        15.914629
51        10.29  2.60  Female     No   Sun  Dinner     2        10.841200
57        26.41  1.50  Female     No   Sat  Dinner     2        26.684208
66        16.45  2.47  Female     No  

Once we have created the groups, we can apply functions over them. For instance, we can calculate the average price of the bill for each of the groups.

In [106]:
for name, group in df.groupby(by=['sex', 'size']):
    print(name, group['total_bill'].mean())
    # equivalente a np.mean(group['total_bill'])

('Female', 1) 6.796666666666667
('Female', 2) 15.339482758620687
('Female', 3) 21.54714285714286
('Female', 4) 30.278888888888886
('Female', 5) 29.85
('Female', 6) 28.425
('Male', 1) 8.58
('Male', 2) 17.10408163265306
('Male', 3) 24.28708333333333
('Male', 4) 28.07821428571429
('Male', 5) 30.122500000000002
('Male', 6) 41.235


The previous method is not so handy if we want to incorporate this information into the actual dataframe, so pandas has for this purpose the function `agg`:

In [107]:
df_bill = df.groupby(['sex']).agg({'total_bill': np.mean})
df_bill

total_bill
sex               
Female   18.056897
Male     20.744076

In [108]:
df_bill = df.groupby(['sex', 'size']).agg({'total_bill': np.mean})
df_bill

total_bill
sex    size            
Female 1       6.796667
       2      15.339483
       3      21.547143
       4      30.278889
       5      29.850000
       6      28.425000
Male   1       8.580000
       2      17.104082
       3      24.287083
       4      28.078214
       5      30.122500
       6      41.235000

Using `pivot`, we can pivot this table in order to see the results better.

In [109]:
df_bill.reset_index().pivot(index='sex', columns='size')

total_bill                                                  
size            1          2          3          4        5       6
sex                                                                
Female   6.796667  15.339483  21.547143  30.278889  29.8500  28.425
Male     8.580000  17.104082  24.287083  28.078214  30.1225  41.235

Do you see anything fishy ? In the female row, the average of total_bill is smaller for a table of 6 people than 4, how can that be ?

In [111]:
df[(df['sex']=='Female')&(df['size']==6)]

total_bill  tip     sex smoker   day   time  size  total_bill_rand
125       29.80  4.2  Female     No  Thur  Lunch     6        29.354890
143       27.05  5.0  Female     No  Thur  Lunch     6        26.226988

In [112]:
df[(df['sex']=='Female')&(df['size']==4)]

total_bill   tip     sex smoker   day    time  size  total_bill_rand
4         24.59  3.61  Female     No   Sun  Dinner     4        23.437885
11        35.26  5.00  Female     No   Sun  Dinner     4        36.013951
33        20.69  2.45  Female     No   Sat  Dinner     4        21.351903
52        34.81  5.20  Female     No   Sun  Dinner     4        36.443165
85        34.83  5.17  Female     No  Thur   Lunch     4        35.231072
119       24.08  2.92  Female     No  Thur   Lunch     4        23.675199
157       25.00  3.75  Female     No   Sun  Dinner     4        26.142584
197       43.11  5.00  Female    Yes  Thur   Lunch     4        45.104112
219       30.14  3.09  Female    Yes   Sat  Dinner     4        27.861675

A possible explanation is that the table of 6 has only 2 cases, which happen to be cheaper meals.

We can pass more than one argument to `agg` if we want to calculate more columns by applying functions on the groups. With this syntax, agg will apply the functions `mean` and `std` to the whole dataframe (where we previously chose the columns total_bill and tip)

In [113]:
df_bill_tips = df.groupby(['sex'])['total_bill', 'tip'].agg([np.mean, np.std])

In [114]:
df_bill_tips

total_bill                 tip          
             mean       std      mean       std
sex                                            
Female  18.056897  8.009209  2.833448  1.159495
Male    20.744076  9.246469  3.089618  1.489102

## Exercise 3
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 

In the tips table select all rows for which the value of `total_bill` is higher that the mean + 1 std deviation for all the meals on the same week day.

How many rows does this dataframe have ?

**Hint:** 

1. Using `groupby` create first a small table whose rows are the days of the week (index is `day`) and whose columns are the mean and std deviation of the total_bill for all meals on that day.

2. Reset the index on that dataframe to turn `day`into an ordinary column.

3. Perform a `merge` of this table with the original table using as key column the `day`, so that each observation in the original table has its corresponding value of avg+std for each day of the week.


In [116]:
df_tips = df.groupby(['day'])['total_bill'].agg([np.mean, np.std])
df_tips.head()

mean       std
day                      
Fri   17.151579  8.302660
Sat   20.441379  9.480419
Sun   21.410000  8.832122
Thur  17.682742  7.886170

In [117]:
df_tips = df_tips.reset_index()


In [118]:
df_tips

day       mean       std
0   Fri  17.151579  8.302660
1   Sat  20.441379  9.480419
2   Sun  21.410000  8.832122
3  Thur  17.682742  7.886170

In [119]:
df_augmented= pd.merge(df, df_tips, how='inner', left_on='day', right_on='day')
df_augmented.head()

total_bill   tip     sex smoker  day    time  size  total_bill_rand   mean  \
0       16.99  1.01  Female     No  Sun  Dinner     2        19.229865  21.41   
1       10.34  1.66    Male     No  Sun  Dinner     3        11.235986  21.41   
2       21.01  3.50    Male     No  Sun  Dinner     3        21.413654  21.41   
3       23.68  3.31    Male     No  Sun  Dinner     2        24.105713  21.41   
4       24.59  3.61  Female     No  Sun  Dinner     4        23.437885  21.41   

        std  
0  8.832122  
1  8.832122  
2  8.832122  
3  8.832122  
4  8.832122

In [124]:
mask=df_augmented['total_bill'] > df_augmented['mean'] + df_augmented['std']
print('The required dataframe has '+str(len(df_augmented[mask]))+ ' rows.')

The required dataframe has 38 rows.


## Exercise 4
<img src="https://www.shareicon.net/data/256x256/2016/06/09/778169_game_512x512.png" alt="Exercise" width="50"/> 

The database `income` can be used to predict if a worker's salary is above or below 50.000 $. For each person, the database contains the following information:
* age 
* education
* marital.status
* relationship
* race
* sex
* hourspeerweek
* nativecountry
* income (whether it is >=50K or <50K)

Load and read the dataset, and eliminate all rows that have some missing value. You can use the method `dropna`. Also, to make things simpler, substitute the values of the `Income` column by 0s and 1s (0 if the salary is smaller than 50K) and 1 otherwise.

In [134]:
dir='data/income.csv'
#dir='/content/Curso-IE/Week_1/data/income.csv'
inc = pd.read_csv(dir)
len(inc)

32561

In [130]:
inc.describe(include='all')

age education      marital.status relationship   race    sex  \
count   32561.000000     32561               32561        32561  32561  32561   
unique           NaN        16                   7            6      5      2   
top              NaN   HS-grad  Married-civ-spouse      Husband  White   Male   
freq             NaN     10501               14976        13193  27816  21790   
mean       38.581647       NaN                 NaN          NaN    NaN    NaN   
std        13.640433       NaN                 NaN          NaN    NaN    NaN   
min        17.000000       NaN                 NaN          NaN    NaN    NaN   
25%        28.000000       NaN                 NaN          NaN    NaN    NaN   
50%        37.000000       NaN                 NaN          NaN    NaN    NaN   
75%        48.000000       NaN                 NaN          NaN    NaN    NaN   
max        90.000000       NaN                 NaN          NaN    NaN    NaN   

        hoursperweek  nativecountry income  
count   32561.000000          31978  32561  
unique           NaN             41      2  
top              NaN  United-States  <=50K  
freq             NaN          29170  24720  
mean       40.437456            NaN    NaN  
std        12.347429            NaN    NaN  
min         1.000000            NaN    NaN  
25%        40.000000            NaN    NaN  
50%        40.000000            NaN    NaN  
75%        45.000000            NaN    NaN  
max        99.000000            NaN    NaN

We see that there are some missing values, let's remove all thpse rows that have some missing values.

In [135]:
inc.dropna(inplace=True)
inc.loc[inc.income == '<=50K', "income"] = 0
inc.loc[inc.income == '>50K', "income"] = 1
inc.head(20)

age     education         marital.status   relationship  \
0    39     Bachelors          Never-married  Not-in-family   
1    50     Bachelors     Married-civ-spouse        Husband   
2    38       HS-grad               Divorced  Not-in-family   
3    53          11th     Married-civ-spouse        Husband   
4    28     Bachelors     Married-civ-spouse           Wife   
5    37       Masters     Married-civ-spouse           Wife   
6    49           9th  Married-spouse-absent  Not-in-family   
7    52       HS-grad     Married-civ-spouse        Husband   
8    31       Masters          Never-married  Not-in-family   
9    42     Bachelors     Married-civ-spouse        Husband   
10   37  Some-college     Married-civ-spouse        Husband   
11   30     Bachelors     Married-civ-spouse        Husband   
12   23     Bachelors          Never-married      Own-child   
13   32    Assoc-acdm          Never-married  Not-in-family   
15   34       7th-8th     Married-civ-spouse        Husband   
16   25       HS-grad          Never-married      Own-child   
17   32       HS-grad          Never-married      Unmarried   
18   38          11th     Married-civ-spouse        Husband   
19   43       Masters               Divorced      Unmarried   
20   40     Doctorate     Married-civ-spouse        Husband   

                  race     sex  hoursperweek  nativecountry  income  
0                White    Male            40  United-States       0  
1                White    Male            13  United-States       0  
2                White    Male            40  United-States       0  
3                Black    Male            40  United-States       0  
4                Black  Female            40           Cuba       0  
5                White  Female            40  United-States       0  
6                Black  Female            16        Jamaica       0  
7                White    Male            45  United-States       1  
8                White  Female            50  United-States       1  
9                White    Male            40  United-States       1  
10               Black    Male            80  United-States       1  
11  Asian-Pac-Islander    Male            40          India       1  
12               White  Female            30  United-States       0  
13               Black    Male            50  United-States       0  
15  Amer-Indian-Eskimo    Male            45         Mexico       0  
16               White    Male            35  United-States       0  
17               White    Male            40  United-States       0  
18               White    Male            50  United-States       0  
19               White  Female            45  United-States       1  
20               White    Male            60  United-States       1

In [137]:
print(len(inc))

31978


Try to come up with some simple rule to predict the variable `income`, and store your prediction on a new variable `income_pred`. You can check how good your prediction is by calling the function precision, that takes two vectors and computes the fraction of correct attributions to total number of predictions.

In [138]:
def precision(income, income_pred):
    ac = sum(income == income_pred)*100/(len(income))
    return "Precision: " + str(ac) + "%"


Let us start with a very simple rule:

1. We predict that only those with a PhD earn more than 50K, and those without it, earn less than 50k

In [139]:
inc["income_pred"] = 0

condition = (inc.education.isin(["Doctorate"])) 
inc.loc[condition, "income_pred"] = 1
precision(inc.income, inc.income_pred)

'Precision: 76.53699418350115%'

Play around combining logical conditions and try to improve that precision.